File: src/023/p023.py
from src.utils.py.math import proper_divisors


def solve() -> int:
    abundant = set()
    non_abundant_sum = 0
    for num in range(1, 28124):
        if sum(proper_divisors(num)) > num:
            abundant.add(num)

        # See if the number can be written as sum of existing abundant numbers
        if not any((num - a in abundant) for a in abundant):
            non_abundant_sum += num
    return non_abundant_sum

File: src/P023/p023.ts
import { sumOfDivisors } from "../utils";

const isAbundant = (n: number): boolean => {
  return sumOfDivisors(n) - n > n;
};

const isAbundantSum = (n: number, abundant: Set<number>): boolean => {
  for (const a of abundant) {
    if (a > n / 2) {
      return false;
    }

    if (abundant.has(n - a)) {
      return true;
    }
  }
  return false;
};

export const solve = (): string => {
  const abundant = new Set<number>();
  let nonAbundantSum = 0;
  for (let i = 1; i <= 28123; i++) {
    if (isAbundant(i)) {
      abundant.add(i);
    }

    if (!isAbundantSum(i, abundant)) {
      nonAbundantSum += i;
    }
  }
  return nonAbundantSum.toString();
};
